### PR TITLE
style(utils): add type hints to utils

### PR DESCRIPTION
**Pull request by P403n1x87**
## Checklist
- [ ] Added to the correct milestone.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Library documentation is updated.
- [ ] [Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).

### CODE REVIEW

File: ddtrace/context.py [hunk 1]
```diff
@@ -34,7 +34,7 @@ class Context(object):
     """
 
     _partial_flush_enabled = asbool(get_env("tracer", "partial_flush_enabled", default=False))
-    _partial_flush_min_spans = int(get_env("tracer", "partial_flush_min_spans", default=500))
+    _partial_flush_min_spans = int(get_env("tracer", "partial_flush_min_spans", default=500))  # type: ignore[arg-type]
```
[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/context.py:37]: It has to do with the fact that `get_env` pulls `default` from `kwargs` so type inference gets mighty confused:

~~~
error: Argument 1 to "int" has incompatible type "Union[str, SupportsInt, _SupportsIndex, _SupportsTrunc, None]"; expected "Union[str, SupportsInt, _SupportsIndex, _SupportsTrunc]"
~~~
So here `int` is never getting a `None` because `default` is passed in to `get_env`

File: ddtrace/profiling/collector/memalloc.py [hunk 2]
```diff
@@ -58,7 +58,9 @@ class MemoryCollector(collector.PeriodicCollector):
     heap_sample_size = attr.ib(
         factory=attr_utils.from_env("DD_PROFILING_HEAP_SAMPLE_SIZE", _DEFAULT_HEAP_SAMPLE_SIZE, int)
     )
-    ignore_profiler = attr.ib(factory=attr_utils.from_env("DD_PROFILING_IGNORE_PROFILER", True, formats.asbool))
+    ignore_profiler = attr.ib(
+        factory=attr_utils.from_env("DD_PROFILING_IGNORE_PROFILER", True, formats.asbool)  # type: ignore[arg-type]
```
[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/profiling/collector/memalloc.py:62]: This is due to the fact that `formats.asbool` is not an instance of `Type`, like in all other cases (e.g. `float`)
~~~
error: Argument 3 to "from_env" has incompatible type "Callable[[Union[str, bool, None]], bool]"; expected "Type[bool]"  [arg-type]
~~~
This could be fixed by allowing both `Type` and `Callable` in `from_env` I think

File: ddtrace/utils/attr.py [hunk 3]
```diff
@@ -1,5 +1,12 @@
...
+from typing import TypeVar
+
+
+T = TypeVar("T")
 
 
 def from_env(name, default, value_type):
-    return lambda: value_type(os.environ.get(name, default))
+    # type: (str, T, Type[T]) -> Callable[[], T]
+    return lambda: value_type(os.environ.get(name, default))  # type: ignore
```
[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/utils/attr.py:12]:
~~~
Too many arguments for "object"  [call-arg]
~~~

File: ddtrace/context.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jd) [reply ddtrace/context.py:37]: Ok.
It might be better to pass the type to `get_env` rather than ignoring the typing info here.

File: ddtrace/profiling/collector/memalloc.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jd) [reply ddtrace/profiling/collector/memalloc.py:62]: Yes, sounds like a `from_env` typing issue we should fix

File: ddtrace/utils/attr.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jd) [reply ddtrace/utils/attr.py:12]: what… does… that… mean? lol
is rewriting that as a `def` helping?

File: ddtrace/context.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/context.py:37]: That's fine by me. @brettlangdon any thoughts?

[the remark this answers is not in the document]
Comment (brettlangdon) [reply ddtrace/context.py:37]: I wonder if we have `get_env` always return `Optional[str]` and `default` must be `Optional[str]`, we'll be wasting some time converting a default of `false` with `asbool` but it'll give us a consistent experience with type usage on `get_env`

... or we use a generic and have the response be `Optional[T]` where `default` is also `Optional[T]` or... something, /shrug

[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/context.py:37]: It looks like whatever I try I end up with having to make changes that perhaps are not worth these few type hints

File: ddtrace/utils/wrappers.py [hunk 4]
```diff
@@ -1,25 +1,46 @@
...
 
+if TYPE_CHECKING:
+    import ddtrace
+
+
+F = TypeVar("F", bound=Callable[..., Any])
+
+
 def iswrapped(obj, attr=None):
+    # type: (Any, str) -> bool
```
[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/utils/wrappers.py:20]: good catch!

File: ddtrace/utils/deprecation.py [hunk 5]
```diff
@@ -1,33 +1,44 @@
...
+from typing import Callable
+from typing import Dict
+from typing import Optional
+from typing import Tuple
 import warnings
 
 from ddtrace.vendor import debtcollector
 
 
+F = Callable[[Tuple[Any], Dict[str, Any]], Any]
```
Comment (Yun-Kim) [inline ddtrace/utils/deprecation.py:13]: Is this used anywhere?

File: ddtrace/utils/importlib.py [hunk 6]
```diff
@@ -15,19 +22,23 @@ def __init__(self, modules):
                 self._missing_modules.append(module)
 
     def __enter__(self):
+        # type: () -> List[str]
         return self._missing_modules
 
     def __exit__(self, exc_type, exc_value, traceback):
-        return False
+        # type: (Optional[Type[BaseException]], Optional[BaseException], Optional[TracebackType]) -> None
+        return
```
Comment (Yun-Kim) [inline ddtrace/utils/importlib.py:30]: Just curious, why did you remove the `return False` / what effect does this have?

**Review by Yun-Kim — COMMENTED** [review]
2 quick questions, looks good otherwise 😄

File: ddtrace/utils/importlib.py (hunk 6 above)
Comment (P403n1x87) [reply ddtrace/utils/importlib.py:30]: `mypy` wanted me to use `Literal` which lives in `mypy_extensions`. Alternatively it suggested changing the return type to `None`. I've opted for the latter to avoid a dependency on `mypy_extensions`

File: ddtrace/utils/deprecation.py (hunk 5 above)
Comment (P403n1x87) [reply ddtrace/utils/deprecation.py:13]: Ah, Good catch. Not anymore!

**Review by Yun-Kim — DISMISSED** [review]

**Review by jd — DISMISSED** [review]

**Review by Yun-Kim — DISMISSED** [review]

File: ddtrace/utils/attr.py [hunk 7]
```diff
@@ -1,5 +1,16 @@
...
+
+
+T = TypeVar("T")
 
 
 def from_env(name, default, value_type):
-    return lambda: value_type(os.environ.get(name, default))
+    # type: (str, T, Union[Callable[[Union[str, T, None]], T], Type[T]]) -> Callable[[], T]
+    def _():
+        value_type(os.environ.get(name, default))
```
Comment (Kyle-Verhoog) [inline ddtrace/utils/attr.py:14]:
```suggestion
        return value_type(os.environ.get(name, default))
```

Comment (P403n1x87) [reply ddtrace/utils/attr.py:14]: 🤦 🤦

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by Yun-Kim — APPROVED** [review]
